lib-backend-wasm/src/func.rs

```diff
@@ -1292,15 +1501,20 @@ fn encode_appl<H: HeapManager>(
         // net wasm stack: [] -> [tableidx]
         expr_builder.local_get(mutctx.wasm_local_slice(localidx_func)[0]);
 
-        // todo!(For optimisation, heap_encode_prologue_epilogue should only be called if the callee might allocate)
-        // Note: encode_args_to_call_function should be *before* encode_local_roots_prologue, since the args themselves might make function calls.
-        if true {
+        // When ctx.options.tail_call, use return_call_appl to call with return_indirect_call
+        // opcode in case of is_tail and call_indirect in case of !is_tail.
+        if ctx.options.wasm_tail_call {
+            // net wasm: [<Func details>] -> [ir_return_type]
+            encode_return_call_appl(ctx, mutctx, expr_builder, is_tail);
+        } else {
+            // todo!(For optimisation, heap_encode_prologue_epilogue should only be called if the callee might allocate)
+            // Note: encode_args_to_call_function should be *before* encode_local_roots_prologue, since the args themselves might make function calls.
             // This function might allocate memory, so we need to store the locals in the gc_roots stack first.
 
             // call the function with gc prologue and epilogue
+            // net wasm: [<Func details>] -> [i32(is_tail)]
             mutctx.heap_encode_prologue_epilogue(ctx.heap, expr_builder, |mutctx, expr_builder| {
```

Comment:
I think we can make an optimisation here by not calling `mutctx.heap_encode_prologue_epilogue` more than once in the entire trampoline.  Something like this:

```rust
mutctx.heap_encode_prologue_epilogue(..., {
    expr_builder.call_indirect(...);
    encode_trampoline(...);
});
```
and `encode_tail_calling_conv()` (called by `encode_trampoline`) should not do `mutctx.heap_encode_prologue_epilogue` again.